### PR TITLE
Add support for all Guest Customization options in resource and data source `vcd_vapp_vm`

### LINKED ISSUE #52: Guest customization - cannot reset root password
### Terraform Version
v0.10.2

### Provider vCD Version
version = "~> 1.0"

### Affected Resource(s)
Please list the resources as a list, for example:
- vcd_vapp

### Terraform Configuration Files
```hcl
resource "vcd_vapp" "server" {
  name              = "TERRA-TEST"
  catalog_name      = "Linux"
  template_name     = "TEMP-CentOS-7.3"
  memory            = 1024
  cpus              = 1
  network_name      = "${var.vcd_number}_NETWORK_01"
  storage_profile   = "SILVER"
  ip                = "${cidrhost(var.internal_net, 110)}"
}
```

### Expected Behavior
vApp should be deployed and VM should be customized. The template has checked "Enable guest customization" and also "Allow local administrator password", "Auto generate password" and "Require administrator to change the password on first login", so the "Guest OS Customization" tab in the new VM should have the same configuration. Root password should be generated and listed on the tab. 

image

### Actual Behavior
The new deployed VM has unchecked mentioned checkboxes (excluding "Enable guest customization") and the root password is not changed.

image

### Steps to Reproduction
1. Run terraform apply
2. Check "Guest OS Customization" tab in the newly deployed VM.
3. Check whether "Allow local administrator password", "Auto generate password" and "Require administrator to change the password on first login" check boxes are checked.

### Important Factoids
Running on vCloud Director 8.10 - Linux OS.

### LINKED ISSUE #303: Add support for `Change SID`
Add support for `Change SID`

### LINKED ISSUE #427: VM guest customization force=false crashes with interface {} is nil
Hi folks, trying to deploy a VM to a vapp and terraform crashes on this line with the error below.

https://github.com/terraform-providers/terraform-provider-vcd/blob/759be139ecfadaf78198876d0392530d50f40391/vcd/resource_vcd_vapp_vm.go#L1343

Here's config

> resource "vcd_vapp_vm" "k8s-masters" {
  count         = var.k8s_master_count
  vapp_name     = var.vapp_name
  name          = "centos-7-test"
  computer_name = "centos-7-test"
  catalog_name  = var.catalog_name
  template_name = var.template_name
  memory        = 4096
  cpus          = 8
  cpu_cores     = 2
customization {
    force = false
  }
network {
    type               = "org"
    name               = var.network_name
    ip_allocation_mode = "POOL"
    is_primary         = true
  }
}
terraform versions: 
Terraform v0.12.18
+ provider.random v2.2.1
+ provider.vcd v2.6.0

> 2020-01-08T22:04:59.617Z [DEBUG] plugin.terraform-provider-vcd_v2.6.0_x4: github.com/terraform-providers/terraform-provider-vcd/vcd.isForcedCustomization(...)
2020-01-08T22:04:59.617Z [DEBUG] plugin.terraform-provider-vcd_v2.6.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-vcd/vcd/resource_vcd_vapp_vm.go:1343

Any thoughts?

### PR DESCRIPTION
**Pull request by Didainius**
Closes #303, closes #52, closes #427

This PR adds support for all Guest customization options available in the UI and adds the following fields to `customization` block:
* enabled
* change_sid
* allow_local_admin_password
* must_change_password_on_first_login
* auto_generate_password
* admin_password
* number_of_auto_logons
* join_domain
* join_org_domain
* join_domain_name
* join_domain_user
* join_domain_password
* join_domain_account_ou
* initscript (the root component `initscript` is deprecated and moved to `customization` block to reflect UI behavior and API position)

All the above fields are _computed_ and _optional_. This means that there are no breaking changes. User is free to set only some of the fields and the others will not be overridden as they may be coming from catalog item itself.

**Note** it is pretty much hard to cover behavior in tests because majority of these fields require a Windows machine, which is too big to include into testing routine.

##### Bugs covered in this PR:
* `resource/vcd_vapp_vm` when a single value of  customization.0.force=false was specified in the `customization` block - it would crash with interface {} is nil error (was previously raised as a separate PR https://github.com/terraform-providers/terraform-provider-vcd/pull/456, but because it was not merged and these PRs go over each other I have closed that one)
* `resource/vcd_vapp_vm` `customization.0.force=true` could have skipped "Forced customization" on each apply

##### Extra features:
* Adds `make test-upgrade-prepare` with similar behavior to `make test-binary-prepare` to only generate files and not trigger test run itself

##### Tests have passed on the following:
* 9.5 and 10.0 (acceptance tests)
* 9.5 (binary tests)
* 9.5 (upgrade tests)

### CODE REVIEW

File: website/docs/r/vapp_vm.html.markdown [hunk 1]
```diff
@@ -288,14 +290,44 @@ Changes are ignored on update. This part isn't reread on refresh. To manage inte
...
+
 * `force` (Optional) **Warning.** `true` value will cause the VM to reboot on every `apply` operation.
 This field works as a flag and triggers force customization when `true` during an update 
 (`terraform apply`) every time. It never complains about a change in statefile. Can be used when guest customization
 is needed after VM configuration (e.g. NIC change, customization options change, etc.) and then set back to `false`.
 **Note.** It will not have effect when `power_on` field is set to `false`. See [example workflow below](#example-forced-customization-workflow).
+* `enabled` (Optional; *v2.7+*) `true` will enable guest customization which may occur on first boot or if the `force` flag is used.
+This option should be selected for **Power on and Force re-customization to work**. For backwards compatibility it is
+enabled by default when deprecated field `initscript` is used.
+* `change_sid` (Optional; *v2.7+*) Allows to change SID (security identifier). Only applicable for Windows operating systems.
```
Comment (lvirbalas) [inline website/docs/r/vapp_vm.html.markdown]: > Only applicable for Windows operating systems.

This is a useful comment. Are there more options below which should have it too?

Comment (Didainius) [reply website/docs/r/vapp_vm.html.markdown]: The customization is pretty much "mystified". Many of the fields may work or not work properly based on the guest operating system. What I did find is that `change_sid` is mentioned exclusively in vCD documentation - https://docs.vmware.com/en/vCloud-Director/9.7/com.vmware.vcloud.user.doc/GUID-BB682E4D-DCD7-4936-A665-0B0FBD6F0EB5.html

File: website/docs/r/vapp_vm.html.markdown [hunk 2]
```diff
@@ -288,14 +290,44 @@ Changes are ignored on update. This part isn't reread on refresh. To manage inte
...
+* `number_of_auto_logons` (Optional; *v2.7+*) Number of times to log on automatically. `0` means disabled.
+* `join_domain` (Optional; *v2.7+*) Enable this VM to join a domain.
+* `join_org_domain` (Optional; *v2.7+*) Set to `true` to use organization's domain.
+* `join_domain_name` (Optional; *v2.7+*) Set the domain name to override organization's domain name.
+* `join_domain_user` (Optional; *v2.7+*) User to be used for domain join.
+* `join_domain_password` (Optional; *v2.7+*) Password to be used for domain join.
+* `join_domain_account_ou` (Optional; *v2.7+*) Organizational unit to be used for domain join.
+* `initscript` (Optional; *v2.7+*) Provide initscript to be executed when customization is applied.
 
 ## Example forced customization workflow
```
Comment (lvirbalas) [inline website/docs/r/vapp_vm.html.markdown]:
```suggestion
## Example of a Forced Customization Workflow
```

Also, nit: the `// ...` in the examples below are out of indent and missing new line after them.

Screenshot 2020-03-03 at 16 19 20

Comment (Didainius) [reply website/docs/r/vapp_vm.html.markdown]: Changed  from `//` to `#` (the real HCL comment symbol).

File: vcd/resource_vcd_vapp_vm.go [hunk 3]
```diff
@@ -1805,3 +1862,161 @@ func resourceVcdVappVmImport(d *schema.ResourceData, meta interface{}) ([]*schem
...
+		customizationSection.ComputerName = computerName
+	}
+
+	// Process parameters from 'customization' block
+	customizationInterface := d.Get("customization")
+	customizationSlice := customizationInterface.([]interface{})
+	if len(customizationSlice) == 1 {
+		cust := customizationSlice[0]
+		if cust != nil {
+			// customBlock := cust.(map[string]interface{})
```
Comment (vbauzysvmware) [inline vcd/resource_vcd_vapp_vm.go]: do we still need this comment?

File: vcd/resource_vcd_vapp_vm.go [hunk 4]
```diff
@@ -1805,3 +1862,161 @@ func resourceVcdVappVmImport(d *schema.ResourceData, meta interface{}) ([]*schem
...
+
+			// Only react to "enabled" field when legacy `initscript` is not specified. Legacy behavior is such that when `initscript`
+			// is specified - guest customization is enabled by default therefore we ignore "enabled" field
+			if _, isSetDeprecatedInitScript := d.GetOk("initscript"); !isSetDeprecatedInitScript {
+				// if enabled, isEnabled := customBlock["enabled"]; isEnabled {
```
Comment (vbauzysvmware) [inline vcd/resource_vcd_vapp_vm.go]: do we still need this comment?

File: vcd/resource_vcd_vapp_vm.go (hunk 3 above)
Comment (Didainius) [reply vcd/resource_vcd_vapp_vm.go]: Nope. removed.

File: vcd/resource_vcd_vapp_vm.go (hunk 4 above)
Comment (Didainius) [reply vcd/resource_vcd_vapp_vm.go]: Nope. removed.

**Pull request merged**